File: caesar_xor.js
// caesar_xor.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const readline = require('readline');

// ANSI colors
const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

// Caesar functions
function caesarEncrypt(text, shift) {
    shift = ((shift % 26) + 26) % 26;
    return text.replace(/[A-Za-z]/g, (ch) => {
        const base = ch >= 'a' ? 'a'.charCodeAt(0) : 'A'.charCodeAt(0);
        const code = ch.charCodeAt(0);
        return String.fromCharCode(((code - base + shift) % 26) + base);
    });
}

function caesarDecrypt(text, shift) {
    return caesarEncrypt(text, -shift);
}

function caesarBruteforce(text) {
    const candidates = [];
    for (let s = 0; s < 26; s++) {
        const dec = caesarEncrypt(text, -s);
        let score = 0;
        for (const ch of dec.toLowerCase()) {
            if ('etaoinshrdlu'.includes(ch)) score++;
        }
        candidates.push({ shift: s, dec, score });
    }
    candidates.sort((a, b) => b.score - a.score);
    return candidates;
}

// XOR (symmetrical)
function xorCipher(text, key) {
    if (!key) throw new Error('Key cannot be empty');
    const keyBytes = Buffer.from(key, 'utf8');
    const textBytes = Buffer.from(text, 'utf8');
    const result = Buffer.alloc(textBytes.length);
    for (let i = 0; i < textBytes.length; i++) {
        result[i] = textBytes[i] ^ keyBytes[i % keyBytes.length];
    }
    return result.toString('utf8');
}

function readFile(filename) {
    return fs.readFileSync(filename, 'utf8');
}

function writeFile(filename, content) {
    fs.writeFileSync(filename, content, 'utf8');
}

function parseArgs() {
    const args = process.argv.slice(2);
    const opts = {
        encrypt: false,
        decrypt: false,
        caesar: false,
        xor: false,
        key: '',
        text: '',
        input: '',
        output: '',
        bruteforce: false,
    };
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        switch (arg) {
            case '-e':
            case '--encrypt': opts.encrypt = true; break;
            case '-d':
            case '--decrypt': opts.decrypt = true; break;
            case '-c':
            case '--caesar': opts.caesar = true; break;
            case '-x':
            case '--xor': opts.xor = true; break;
            case '-k':
            case '--key': opts.key = args[++i]; break;
            case '-t':
            case '--text': opts.text = args[++i]; break;
            case '-i':
            case '--input': opts.input = args[++i]; break;
            case '-o':
            case '--output': opts.output = args[++i]; break;
            case '-b':
            case '--bruteforce': opts.bruteforce = true; break;
            case '-h':
            case '--help':
                console.log(`Usage: node caesar_xor.js [options]
Options:
  -e, --encrypt       Encrypt mode (default)
  -d, --decrypt       Decrypt mode
  -c, --caesar        Use Caesar cipher
  -x, --xor           Use XOR cipher
  -k, --key <key>     Key: shift (int) for Caesar or string for XOR
  -t, --text <text>   Input text
  -i, --input <file>  Input file
  -o, --output <file> Output file
  -b, --bruteforce    Bruteforce Caesar (ignores key)
  -h, --help          Show this help`);
                process.exit(0);
            default:
                console.error(colorize(`Unknown option: ${arg}`, 'red'));
                process.exit(1);
        }
    }
    return opts;
}

async function main() {
    const opts = parseArgs();

    if (opts.decrypt && opts.encrypt) {
        console.error(colorize('Error: cannot use both -e and -d', 'red'));
        process.exit(1);
    }
    const mode = opts.decrypt ? 'decrypt' : 'encrypt';

    if (opts.caesar && opts.xor) {
        console.error(colorize('Error: choose only one cipher (-c or -x)', 'red'));
        process.exit(1);
    }
    if (!opts.caesar && !opts.xor) {
        console.error(colorize('Error: specify cipher (-c or -x)', 'red'));
        process.exit(1);
    }

    // Read input
    let inputText = '';
    if (opts.text) {
        inputText = opts.text;
    } else if (opts.input) {
        try {
            inputText = readFile(opts.input);
        } catch (err) {
            console.error(colorize(`Error reading file: ${err.message}`, 'red'));
            process.exit(1);
        }
    } else {
        // read from stdin
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: false
        });
        const lines = [];
        for await (const line of rl) {
            lines.push(line);
        }
        inputText = lines.join('\n');
        if (!inputText && process.stdin.isTTY) {
            console.error(colorize('Error: no input text', 'red'));
            process.exit(1);
        }
    }

    let result = '';
    try {
        if (opts.caesar) {
            if (opts.bruteforce) {
                console.log(colorize('Bruteforce Caesar mode:', 'blue'));
                const candidates = caesarBruteforce(inputText);
                for (let i = 0; i < Math.min(5, candidates.length); i++) {
                    const c = candidates[i];
                    console.log(`Shift ${c.shift.toString().padStart(2)} (score ${c.score}): ${colorize(c.dec, 'green')}`);
                }
                const best = candidates[0];
                console.log(colorize(`\nMost likely shift: ${best.shift}`, 'yellow'));
                console.log(colorize(`Result: ${best.dec}`, 'green'));
                result = best.dec;
            } else {
                const shift = parseInt(opts.key);
                if (isNaN(shift)) {
                    console.error(colorize('Key for Caesar must be an integer', 'red'));
                    process.exit(1);
                }
                if (mode === 'encrypt') {
                    result = caesarEncrypt(inputText, shift);
                } else {
                    result = caesarDecrypt(inputText, shift);
                }
            }
        } else { // XOR
            if (!opts.key) {
                console.error(colorize('Key for XOR cannot be empty', 'red'));
                process.exit(1);
            }
            result = xorCipher(inputText, opts.key);
        }
    } catch (err) {
        console.error(colorize(`Error: ${err.message}`, 'red'));
        process.exit(1);
    }

    if (opts.output) {
        try {
            writeFile(opts.output, result);
            console.log(colorize(`Result written to ${opts.output}`, 'green'));
        } catch (err) {
            console.error(colorize(`Error writing file: ${err.message}`, 'red'));
            process.exit(1);
        }
    } else {
        process.stdout.write(result);
    }
}

main().catch(err => {
    console.error(colorize(`Unhandled error: ${err.message}`, 'red'));
    process.exit(1);
});

File: caesar_xor.rb
#!/usr/bin/env ruby
# caesar_xor.rb
# encoding: UTF-8

require 'optparse'
require 'io/console'

# ANSI colors
COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

# Caesar encrypt (ASCII letters only for simplicity)
def caesar_encrypt(text, shift)
  shift = shift % 26
  text.tr('A-Za-z', ('A'..'Z').to_a.rotate(shift).join + ('a'..'z').to_a.rotate(shift).join)
end

def caesar_decrypt(text, shift)
  caesar_encrypt(text, -shift)
end

def caesar_bruteforce(text)
  candidates = []
  (0...26).each do |s|
    dec = caesar_encrypt(text, -s)
    score = dec.count('etaoinshrdluETAOINSHRDLU')
    candidates << { shift: s, dec: dec, score: score }
  end
  candidates.sort_by! { |c| -c[:score] }
  candidates
end

# XOR (symmetrical)
def xor_cipher(text, key)
  raise "Key cannot be empty" if key.empty?
  key_bytes = key.bytes
  text.bytes.map.with_index { |b, i| (b ^ key_bytes[i % key_bytes.length]).chr }.join
end

def read_file(filename)
  File.read(filename, encoding: 'UTF-8')
rescue => e
  raise "Cannot read file: #{e.message}"
end

def write_file(filename, content)
  File.write(filename, content, encoding: 'UTF-8')
rescue => e
  raise "Cannot write file: #{e.message}"
end

options = {
  encrypt: false,
  decrypt: false,
  caesar: false,
  xor: false,
  key: nil,
  text: nil,
  input: nil,
  output: nil,
  bruteforce: false
}

OptionParser.new do |opts|
  opts.banner = "Usage: caesar_xor.rb [options]"
  opts.on('-e', '--encrypt', 'Encrypt mode (default)') { options[:encrypt] = true }
  opts.on('-d', '--decrypt', 'Decrypt mode') { options[:decrypt] = true }
  opts.on('-c', '--caesar', 'Use Caesar cipher') { options[:caesar] = true }
  opts.on('-x', '--xor', 'Use XOR cipher') { options[:xor] = true }
  opts.on('-k', '--key KEY', 'Key: shift (int) for Caesar or string for XOR') { |k| options[:key] = k }
  opts.on('-t', '--text TEXT', 'Input text') { |t| options[:text] = t }
  opts.on('-i', '--input FILE', 'Input file') { |f| options[:input] = f }
  opts.on('-o', '--output FILE', 'Output file') { |f| options[:output] = f }
  opts.on('-b', '--bruteforce', 'Bruteforce Caesar (ignores key)') { options[:bruteforce] = true }
  opts.on('-h', '--help', 'Show this help') do
    puts opts
    exit
  end
end.parse!

if options[:decrypt] && options[:encrypt]
  puts colorize("Error: cannot use both -e and -d", :red)
  exit 1
end
mode = options[:decrypt] ? 'decrypt' : 'encrypt'

if options[:caesar] && options[:xor]
  puts colorize("Error: choose only one cipher (-c or -x)", :red)
  exit 1
end
if !options[:caesar] && !options[:xor]
  puts colorize("Error: specify cipher (-c or -x)", :red)
  exit 1
end

# Read input
input_text = nil
if options[:text]
  input_text = options[:text]
elsif options[:input]
  begin
    input_text = read_file(options[:input])
  rescue => e
    puts colorize("Error reading input file: #{e}", :red)
    exit 1
  end
else
  # read from stdin
  if $stdin.tty?
    print colorize("Enter text (end with Ctrl+D): ", :yellow)
  end
  input_text = $stdin.read
  if input_text.nil? || input_text.empty?
    puts colorize("No input text", :red)
    exit 1
  end
end

result = nil
begin
  if options[:caesar]
    if options[:bruteforce]
      puts colorize("Bruteforce Caesar mode:", :blue)
      candidates = caesar_bruteforce(input_text)
      candidates.first(5).each do |c|
        puts "Shift #{c[:shift].to_s.rjust(2)} (score #{c[:score]}): #{colorize(c[:dec], :green)}"
      end
      best = candidates.first
      puts colorize("\nMost likely shift: #{best[:shift]}", :yellow)
      puts colorize("Result: #{best[:dec]}", :green)
      result = best[:dec]
    else
      shift = options[:key].to_i
      if mode == 'encrypt'
        result = caesar_encrypt(input_text, shift)
      else
        result = caesar_decrypt(input_text, shift)
      end
    end
  else # XOR
    raise "Key for XOR cannot be empty" if options[:key].nil? || options[:key].empty?
    result = xor_cipher(input_text, options[:key])
  end
rescue => e
  puts colorize("Error: #{e.message}", :red)
  exit 1
end

if options[:output]
  begin
    write_file(options[:output], result)
    puts colorize("Result written to #{options[:output]}", :green)
  rescue => e
    puts colorize("Error writing output: #{e}", :red)
    exit 1
  end
else
  print result
end
